
## Diabetes chr6_32408 example -- Which component is important?

- The specific context of this variant and disease is described in this google doc:
  - https://docs.google.com/document/d/16GuSasXWX-5qwvKAX5-4VxtrbmsIu9UgrP311_viqQc/edit?usp=sharing
- This notebook would show, 
  1. Given the SNP, identify which genomic bin contains the SNP
  1. Use genomic bin squared cosine score to find the top 3 important components for the genomic bin
  1. Investigate the top component for the genomic bins
    - Use assay contribution scores to see what assays are important for the component
    - Use genomic bin contribution scores to see what other gnomic bins are important for the component
    - Explorer the results of enrichment analysis


In [1]:
% matplotlib inline

import numpy as np
import pandas as pd
import matplotlib, collections, itertools, os, re, textwrap, logging, sys
import matplotlib.pyplot as plt
import matplotlib.gridspec as gridspec
import matplotlib.patches as mpatches
from functools import reduce

from logging.config import dictConfig
from logging import getLogger

dictConfig(dict(
    version = 1,
    formatters = {'f': {'format': '%(asctime)s %(name)-12s %(levelname)-8s %(message)s'}},
    handlers = {
        'h': {'class': 'logging.StreamHandler','formatter': 'f',
              'level': logging.DEBUG}},
    root = {'handlers': ['h'], 'level': logging.DEBUG,},
))

matplotlib.rc('font',**{'size':16, 'family':'sans-serif','sans-serif':['HelveticaNeue', 'Helvetica']})

logger = getLogger('notebook')


In [2]:
repo_dir=os.path.realpath(
    os.path.dirname(os.path.dirname(os.getcwd()))
)


In [3]:
data_dir=os.path.realpath(
    os.path.join(os.path.dirname(os.getcwd()), 'private_data')
)

In [4]:
enrichment_data_dir=os.path.join(repo_dir, 'enrichment', 'private_data')


In [5]:
sys.path.append(os.path.join(repo_dir, 'enrichment', 'src'))
from great import read_great_res_wrapper


In [6]:
metadata_dir=os.path.join(
    repo_dir, 'metadata'
)
metadata = pd.read_table(
    os.path.join(metadata_dir, 'sample_antibody_map_v2_with_metadata.tsv'),
    sep='\t',
)

### Step 1: SNP to genomic bin


In [7]:
genomic_bin_df=pd.read_csv(
    os.path.join(repo_dir, 'enrichment', 'private_data', 'loci_def.bed'),
    names=['chr', 'chromStart', 'chromEnd', 'name'],
    sep='\t'
)

In [8]:
genomic_bin_df[genomic_bin_df['name'] == 'chr6_32408']

chr  chromStart  chromEnd        name
143184  chr6    32408000  32409000  chr6_32408

This means the index of the genomic bin of our interest is 143278

In [9]:
genomic_bin_idx=143184

### Step 2: Which component is important for a given genomic bin -- genomic bin squared cosine score
- Let's write our decomposition as X = UDV' where X is input feature matrix, D is diagonal singular value matrix, U is left singular vector matrix (on assay space), V is right singular vector matrix (on genomic bin space), and `'` denotes the transposition of the matrix.
- Genomic bin squared cosine score is defined as L2-normalized version of the matrix product (VD) so that any given slice for a given genomic bin has Euclidian norm of 1. 
- The interpretation of the score is it represents the relative importance of the component given a genomic bin.
- More formal definition:
  - https://docs.google.com/document/d/1YRuaIvHvjb_6SJwlml1dQDegiGlGbdfz_zN-5bneroE/edit?usp=sharing
 

#### read the decomposed matrices

In [10]:
def read_decomposed_matrix(filename, compression=None):
    if((compression is None) and (len(filename) > 3) and (filename[-3:] == '.gz')):
        compression='gzip'
    df = pd.read_csv(
        os.path.join(data_dir, filename),
        compression=compression
    )
    mat = df.iloc[:, 1:].as_matrix()
    idx = df.iloc[:, 0].as_matrix()
    return mat, idx

In [11]:
d_mat_temp, d_idx = read_decomposed_matrix(os.path.join(data_dir, 'diagonalScore.csv.gz'))
d_vec = d_mat_temp[:, 0]


In [12]:
u_mat, u_idx = read_decomposed_matrix(os.path.join(data_dir, 'uScore.csv.gz'))


In [13]:
v_mat, v_idx = read_decomposed_matrix(os.path.join(data_dir, 'vScore.csv.gz'))


In [14]:
d_vec.shape, u_mat.shape, v_mat.shape, d_idx.shape, u_idx.shape, v_idx.shape

((652,), (652, 652), (379541, 652), (652,), (652,), (379541,))

#### compute matrix products, UD and VD

In [15]:
u_dot_d = np.dot(u_mat, np.diag(d_vec))


In [16]:
v_dot_d = np.dot(v_mat, np.diag(d_vec))


In [17]:
u_dot_d.shape, v_dot_d.shape

((652, 652), (379541, 652))

#### compute normalized matrices
- v_dot_d_find_pcs: genomic bin --> which PC? genomic bin squared contribution score.
- u_dot_d_fine_pcs: assay       --> which PC? assay squared contribution score.
- v_dot_d_find_loci: PC --> which genomic bins? genomic bin contribution score.
- u_dot_d_find_assay: PC --> which assay? assay contribution score

In [18]:
v_dot_d_find_pcs = (v_dot_d ** 2 ) / (np.sum(v_dot_d ** 2, axis = 1)[:,np.newaxis])


In [19]:
u_dot_d_find_pcs = (u_dot_d ** 2 ) / (np.sum(u_dot_d ** 2, axis = 1)[:,np.newaxis])


In [20]:
v_dot_d_find_loci = (v_dot_d ** 2 ) / (np.sum(v_dot_d ** 2, axis = 0)[np.newaxis, :])


In [21]:
u_dot_d_find_assay = (u_dot_d ** 2 ) / (np.sum(u_dot_d ** 2, axis = 0)[np.newaxis, :])


#### let's identify the top 3 important components for the genomic bin chr17_38023 (index: 328517)

In [22]:
np.argsort(-v_dot_d_find_pcs[genomic_bin_idx, :])[:5]

array([ 0,  5,  6, 10, 16])

In [23]:
v_dot_d_find_pcs[genomic_bin_idx, np.argsort(-v_dot_d_find_pcs[genomic_bin_idx, :])[:5]]

array([0.25498096, 0.23539941, 0.15935923, 0.02779045, 0.02525449])

This means PC0 (0-based index) is the most important component for this bin with 25.5% of squared cosine score, PC5 is the second important one with 23.5%, etc ...

### Step 3: investigation of the components

#### PC0 (the top component)

We will investigate 

1. What assays are driving this component?
1. What genomic loci are driving this component?
1.  What are the top hits in the enrichment analysis?

In [24]:
component_idx=0

#### what assays are driving this component?

In [25]:
np.argsort(-u_dot_d_find_assay[:, component_idx])[:5]

array([193, 192,  93, 194,  28])

In [26]:
u_dot_d_find_assay[np.argsort(-u_dot_d_find_assay[:, component_idx])[:5], component_idx]

array([0.49684131, 0.31340371, 0.05643114, 0.01414899, 0.00910366])

In [27]:
metadata.iloc[np.argsort(-u_dot_d_find_assay[:, component_idx])[:5], :]

sample_number antibody Genome assembly Antigen class       Antigen  \
193     SRX106085      NaN            hg19       Histone      H3K27me3   
192     SRX106084      NaN            hg19       Histone      H3K27me3   
93      ERX329704      NaN            hg19  Unclassified  Unclassified   
194     SRX106087      NaN            hg19       Histone      H3K27me3   
28      ERX329639      NaN            hg19  Unclassified  Unclassified   

    Cell type class                 Cell type  \
193           Blood  Lymphoblastoid cell line   
192           Blood  Lymphoblastoid cell line   
93            Blood  Lymphoblastoid cell line   
194           Blood  Lymphoblastoid cell line   
28            Blood  Lymphoblastoid cell line   

                                 Cell type description  \
193  Tissue=blood|Lineage=mesoderm|Description=pare...   
192  Tissue=blood|Lineage=mesoderm|Description=pare...   
93   Tissue=blood|Lineage=mesoderm|Description=pare...   
194  Tissue=blood|Lineage=mesoderm|Description=pare...   
28   Tissue=blood|Lineage=mesoderm|Description=pare...   

              Processing logs  \
193  50863687,47.0,38.0,44910   
192  45470156,35.1,60.9,34413   
93   53432856,37.9,67.7,21958   
194  46755455,53.5,22.7,10811   
28   48478749,83.7,14.8,38430   

                                                 Title   ...   age  treatment  \
193    GSM835575: WBS H3K27me3; Homo sapiens; ChIP-Seq   ...   NaN        NaN   
192    GSM835574: SMS H3K27me3; Homo sapiens; ChIP-Seq   ...   NaN        NaN   
93   Illumina HiSeq 2000 sequencing; Coordinated ef...   ...   NaN        NaN   
194   GSM835577: Ctrl H3K27me3; Homo sapiens; ChIP-Seq   ...   NaN        NaN   
28   Illumina Genome Analyzer IIx sequencing; Coord...   ...   NaN        NaN   

    genotype lab  age.1 health state  cell_type  tissue_type  provider     sex  
193      NaN NaN    NaN          NaN        NaN          NaN       NaN     NaN  
192      NaN NaN    NaN          NaN        NaN          NaN       NaN     NaN  
93       NaN NaN    NaN          NaN    NA11830          NaN       NaN  female  
194      NaN NaN    NaN          NaN        NaN          NaN       NaN     NaN  
28       NaN NaN    NaN          NaN    NA12892          NaN       NaN  female  

[5 rows x 24 columns]

#### what genomic bins are driving this component?

In [28]:
np.argsort(-v_dot_d_find_loci[:, component_idx])[:5]

array([310943, 291154,  64617, 213360, 217066])

In [29]:
v_dot_d_find_loci[np.argsort(-v_dot_d_find_loci[:, component_idx])[:5], component_idx]

array([3.14733015e-05, 3.14360980e-05, 3.14233214e-05, 3.14025649e-05,
       3.14000381e-05])

These genomic bins are important for PC16. Note the genomic bin contribution scores are very small compared to assay contribution score. This is expected becasue of the large number of genomic bins in the whole-genome analysis.

#### where is our loci of interest in this ranking?

In [30]:
np.sum(v_dot_d_find_loci[:, component_idx] >= v_dot_d_find_loci[
    genomic_bin_idx, component_idx])

45603

In [31]:
np.sum(v_dot_d_find_loci[:, component_idx] >= v_dot_d_find_loci[
    genomic_bin_idx, component_idx]) / v_dot_d_find_loci.shape[0]

0.12015302694570547

It's roughly on the top 12 percentile. --> not a strong signal ??


#### results of the enrichment analysis

In [33]:
read_great_res_wrapper(enrichment_data_dir, component_idx, 'HumanPhenotypeOntology').head(10)


# ID                                          Desc         BPval  \
3   HP:0012140  Abnormality of cells of the lymphoid lineage  3.637427e-07   
5   HP:0001888                                   Lymphopenia  5.567840e-07   
11  HP:0001878                              Hemolytic anemia  1.152956e-06   
13  HP:0002917                                Hypomagnesemia  3.079205e-06   
20  HP:0004921          Abnormality of magnesium homeostasis  5.371737e-05   
22  HP:0000121                              Nephrocalcinosis  8.832524e-05   
24  HP:0200114                           Metabolic alkalosis  1.073928e-04   
25  HP:0002643                 Neonatal respiratory distress  1.307868e-04   
27  HP:0000360                                      Tinnitus  1.513635e-04   
28  HP:0001281                                        Tetany  1.785555e-04   

       BFold  
3   2.315639  
5   2.382784  
11  2.008804  
13  3.452539  
20  2.814905  
22  2.448755  
24  4.508408  
25  3.743395  
27  2.679365  
28  2.937869

In [34]:
read_great_res_wrapper(enrichment_data_dir, component_idx, 'GOBiologicalProcess').head(10)


# ID                                            Desc         BPval  \
19  GO:0080111                               DNA demethylation  3.780717e-08   
20  GO:0046133     pyrimidine ribonucleoside catabolic process  4.080519e-08   
26  GO:0006703                   estrogen biosynthetic process  9.438592e-08   
29  GO:0045916    negative regulation of complement activation  1.792544e-07   
35  GO:0033081  regulation of T cell differentiation in thymus  5.234933e-07   
40  GO:0002921  negative regulation of humoral immune response  7.788734e-07   
47  GO:0006369  termination of RNA polymerase II transcription  1.373182e-06   
48  GO:0008334                  histone mRNA metabolic process  1.448484e-06   
66  GO:0006749                   glutathione metabolic process  4.592398e-06   
71  GO:0010955       negative regulation of protein processing  5.448677e-06   

       BFold  
19  3.904828  
20  6.249585  
26  3.843655  
29  4.370029  
35  2.308571  
40  3.955009  
47  2.414571  
48  2.745745  
66  2.208914  
71  2.330956

In [35]:
read_great_res_wrapper(enrichment_data_dir, component_idx, 'MGIPhenotype').head(10)


# ID                                               Desc     BPval  \
26  MP:0001856                                        myocarditis  0.000003   
34  MP:0001870                        salivary gland inflammation  0.000007   
36  MP:0004041  increased susceptibility to kidney reperfusion...  0.000008   
44  MP:0002389         abnormal Peyer's patch follicle morphology  0.000018   
60  MP:0002392      abnormal Peyer's patch T cell area morphology  0.000067   
62  MP:0008862                                   asymmetric snout  0.000070   
68  MP:0003452                  abnormal parotid gland morphology  0.000087   
69  MP:0003407       abnormal central nervous system regeneration  0.000088   
72  MP:0009488           abnormal pancreatic islet cell apoptosis  0.000095   
88  MP:0002565                            delayed circadian phase  0.000161   

       BFold  
26  2.899189  
34  2.334535  
36  4.074445  
44  2.146357  
60  3.541650  
62  2.759692  
68  3.455479  
69  2.634864  
72  3.107736  
88  2.458035

In [36]:
read_great_res_wrapper(enrichment_data_dir, component_idx, 'MGIPhenoSingleKO').head(10)


# ID                                               Desc     BPval  \
24  MP:0001856                                        myocarditis  0.000003   
29  MP:0004041  increased susceptibility to kidney reperfusion...  0.000008   
35  MP:0008126                    increased dendritic cell number  0.000017   
59  MP:0008862                                   asymmetric snout  0.000070   
63  MP:0003452                  abnormal parotid gland morphology  0.000087   
64  MP:0001870                        salivary gland inflammation  0.000092   
65  MP:0003031                                           acidosis  0.000095   
81  MP:0002565                            delayed circadian phase  0.000161   
84  MP:0000786  abnormal embryonic neuroepithelial layer diffe...  0.000176   
86  MP:0002389         abnormal Peyer's patch follicle morphology  0.000194   

       BFold  
24  3.121352  
29  4.074445  
35  2.882200  
59  2.759692  
63  3.455479  
64  2.225581  
65  2.154495  
81  2.458035  
84  2.571348  
86  2.095309

In [37]:
read_great_res_wrapper(enrichment_data_dir, component_idx, 'GOCellularComponent').head(10)


# ID                                               Desc  \
4   GO:0042611                                MHC protein complex   
17  GO:0071556  integral to lumenal side of endoplasmic reticu...   
20  GO:0042612                        MHC class I protein complex   
21  GO:0042613                       MHC class II protein complex   
26  GO:0005761                             mitochondrial ribosome   
28  GO:0012507             ER to Golgi transport vesicle membrane   
32  GO:0005763              mitochondrial small ribosomal subunit   
33  GO:0005689                      U12-type spliceosomal complex   
36  GO:0005683                                           U7 snRNP   
43  GO:0000421                        autophagic vacuole membrane   

           BPval     BFold  
4   1.528314e-09  4.026781  
17  5.172804e-06  3.113123  
20  9.299293e-06  3.610470  
21  2.079056e-05  3.733819  
26  4.195153e-05  2.000761  
28  5.810453e-05  2.100427  
32  1.794133e-04  2.642842  
33  2.281263e-04  2.342155  
36  3.130327e-04  3.936596  
43  6.123051e-04  2.136283

In [38]:
read_great_res_wrapper(enrichment_data_dir, component_idx, 'GOMolecularFunction').head(10)


# ID                                               Desc  \
0   GO:0004303           estradiol 17-beta-dehydrogenase activity   
1   GO:0003823                                    antigen binding   
2   GO:0001848                                 complement binding   
3   GO:0033764  steroid dehydrogenase activity, acting on the ...   
5   GO:0016755  transferase activity, transferring amino-acyl ...   
6   GO:0032395                     MHC class II receptor activity   
8   GO:0042605                            peptide antigen binding   
10  GO:0003746             translation elongation factor activity   
11  GO:0016229                     steroid dehydrogenase activity   
14  GO:0017091                            AU-rich element binding   

           BPval     BFold  
0   5.508480e-09  4.707856  
1   2.894243e-08  2.479716  
2   6.373650e-08  5.244508  
3   3.355615e-07  2.687094  
5   1.664452e-06  2.462163  
6   3.954467e-06  4.956089  
8   5.772647e-06  3.000111  
10  8.282372e-06  2.717842  
11  1.135737e-05  2.281044  
14  1.549784e-05  2.895299

#### PC5 (the second component)


In [39]:
component_idx=5

#### what assays are driving this component?

In [40]:
np.argsort(-u_dot_d_find_assay[:, component_idx])[:5]

array([188, 433,  76, 434, 431])

In [41]:
u_dot_d_find_assay[np.argsort(-u_dot_d_find_assay[:, component_idx])[:5], component_idx]

array([0.1195496 , 0.08228496, 0.05328201, 0.03843322, 0.03278715])

In [42]:
metadata.iloc[np.argsort(-u_dot_d_find_assay[:, component_idx])[:5], :]

sample_number antibody Genome assembly   Antigen class       Antigen  \
188    SRX1027619      NaN            hg19  No description           NaN   
433     SRX651491      NaN            hg19         Histone       H3K4me3   
76      ERX329687      NaN            hg19    Unclassified  Unclassified   
434     SRX651492      NaN            hg19         Histone       H3K4me1   
431     SRX651470      NaN            hg19         Histone       H3K4me3   

    Cell type class                 Cell type  \
188           Blood  Lymphoblastoid cell line   
433           Blood  Lymphoblastoid cell line   
76            Blood  Lymphoblastoid cell line   
434           Blood  Lymphoblastoid cell line   
431           Blood  Lymphoblastoid cell line   

                                 Cell type description  \
188  Tissue=blood|Lineage=mesoderm|Description=pare...   
433  Tissue=blood|Lineage=mesoderm|Description=pare...   
76   Tissue=blood|Lineage=mesoderm|Description=pare...   
434  Tissue=blood|Lineage=mesoderm|Description=pare...   
431  Tissue=blood|Lineage=mesoderm|Description=pare...   

                Processing logs  \
188  804451442,97.3,44.1,118575   
433    166047212,93.9,6.4,63296   
76   238094924,94.2,33.9,144042   
434   112405175,94.2,1.6,121976   
431    117212731,95.4,4.0,47422   

                                                 Title  \
188                  ChIP-seq of Homo sapiens: H3K4me3   
433  GSM1435515: LCL19238 H3K4me3; Homo sapiens; Ch...   
76   Illumina HiSeq 2000 sequencing; Coordinated ef...   
434  GSM1435516: LCL19239 H3K4me1; Homo sapiens; Ch...   
431  GSM1435517: LCL19239 H3K4me3; Homo sapiens; Ch...   

              ...           age  treatment genotype lab  age.1 health state  \
188           ...           NaN        NaN      NaN NaN    NaN          NaN   
433           ...           NaN        NaN      NaN NaN    NaN          NaN   
76            ...           NaN        NaN      NaN NaN    NaN          NaN   
434           ...           NaN        NaN      NaN NaN    NaN          NaN   
431           ...           NaN        NaN      NaN NaN    NaN          NaN   

     cell_type  tissue_type  provider                     sex  
188       LCLs        blood   Coriell  pooled male and female  
433        NaN          NaN       NaN                     NaN  
76     NA12878          NaN       NaN                  female  
434        NaN          NaN       NaN                     NaN  
431        NaN          NaN       NaN                     NaN  

[5 rows x 24 columns]

#### what genomic bins are driving this component?

In [43]:
np.argsort(-v_dot_d_find_loci[:, component_idx])[:5]

array([282969, 130825, 187463, 337732, 233484])

In [44]:
v_dot_d_find_loci[np.argsort(-v_dot_d_find_loci[:, component_idx])[:5], component_idx]

array([8.04405926e-05, 6.46164381e-05, 6.45142013e-05, 6.39899652e-05,
       6.37951179e-05])

#### where is our loci of interest in this ranking?

In [45]:
np.sum(v_dot_d_find_loci[:, component_idx] >= v_dot_d_find_loci[genomic_bin_idx, component_idx])

1984

In [46]:
np.sum(v_dot_d_find_loci[:, component_idx] >= v_dot_d_find_loci[genomic_bin_idx, component_idx]) / v_dot_d_find_loci.shape[0]

0.005227366740352162

It's roughly on the top 1 percentile. --> good one


#### results of the enrichment analysis

In [47]:
read_great_res_wrapper(enrichment_data_dir, component_idx, 'HumanPhenotypeOntology').head(10)


# ID                                 Desc         BPval     BFold
0   HP:0002697                    Parietal foramina  4.367679e-08  2.372433
1   HP:0004425                        Flat forehead  1.568965e-06  2.735910
3   HP:0004442            Sagittal craniosynostosis  2.359962e-06  2.733096
4   HP:0002365          Hypoplasia of the brainstem  2.396735e-06  2.031767
5   HP:0010054  Abnormality of the first metatarsal  2.506458e-06  2.574793
7   HP:0006191                   Deep palmar crease  4.189624e-06  2.977180
8   HP:0000557                          Buphthalmos  4.706791e-06  2.287392
12  HP:0009836       Broad distal phalanx of finger  1.739019e-05  2.601329
18  HP:0003741        Congenital muscular dystrophy  3.098136e-05  2.094681
21  HP:0003535          3-Methylglutaconic aciduria  3.861840e-05  2.428799

In [48]:
read_great_res_wrapper(enrichment_data_dir, component_idx, 'GOBiologicalProcess').head(10)


# ID                                               Desc     BPval  \
62   GO:0035089        establishment of apical/basal cell polarity  0.000008   
76   GO:0075733                   intracellular transport of virus  0.000012   
100  GO:0051775                            response to redox state  0.000021   
108  GO:0061162           establishment of monopolar cell polarity  0.000025   
113  GO:0070127  tRNA aminoacylation for mitochondrial protein ...  0.000027   
116  GO:0071624      positive regulation of granulocyte chemotaxis  0.000029   
126  GO:0090023       positive regulation of neutrophil chemotaxis  0.000034   
148  GO:0018401  peptidyl-proline hydroxylation to 4-hydroxy-L-...  0.000056   
161  GO:0001672    regulation of chromatin assembly or disassembly  0.000072   
164  GO:0006999                          nuclear pore organization  0.000077   

        BFold  
62   2.658500  
76   2.073909  
100  2.764178  
108  2.320837  
113  4.115960  
116  2.016287  
126  2.020100  
148  2.373342  
161  3.185283  
164  2.528835

In [49]:
read_great_res_wrapper(enrichment_data_dir, component_idx, 'MGIPhenotype').head(10)


# ID                                        Desc         BPval  \
12  MP:0001771        abnormal circulating magnesium level  1.832349e-07   
19  MP:0003954     abnormal Reichert's membrane morphology  8.388949e-07   
53  MP:0002348      abnormal lymph node medulla morphology  9.682417e-06   
54  MP:0010092       increased circulating magnesium level  1.018919e-05   
59  MP:0009545  abnormal dermis papillary layer morphology  1.604188e-05   
60  MP:0010743                      delayed suture closure  1.697130e-05   
65  MP:0001669                 abnormal glucose absorption  2.245532e-05   
69  MP:0006210                         abnormal orbit size  2.804297e-05   
77  MP:0002050                            pheochromocytoma  3.830129e-05   
79  MP:0000666        decreased prostate gland duct number  4.319200e-05   

       BFold  
12  2.327122  
19  2.004437  
53  2.903855  
54  2.628006  
59  2.274394  
60  2.208579  
65  2.618271  
69  2.181630  
77  2.589431  
79  2.700373

In [50]:
read_great_res_wrapper(enrichment_data_dir, component_idx, 'MGIPhenoSingleKO').head(10)


# ID                                        Desc         BPval  \
5   MP:0012129             failure of blastocyst formation  2.360537e-09   
8   MP:0002663                  failure to form blastocele  3.603307e-09   
23  MP:0003954     abnormal Reichert's membrane morphology  8.388949e-07   
24  MP:0001771        abnormal circulating magnesium level  1.211860e-06   
33  MP:0010743                      delayed suture closure  3.083532e-06   
34  MP:0002050                            pheochromocytoma  3.483130e-06   
44  MP:0002348      abnormal lymph node medulla morphology  9.682417e-06   
48  MP:0009545  abnormal dermis papillary layer morphology  1.604188e-05   
60  MP:0000275                           heart hyperplasia  4.592075e-05   
69  MP:0002031     increased adrenal gland tumor incidence  6.052274e-05   

       BFold  
5   2.005173  
8   2.004262  
23  2.004437  
24  2.284786  
33  2.694599  
34  3.010810  
44  2.903855  
48  2.274394  
60  2.012848  
69  2.408676

In [51]:
read_great_res_wrapper(enrichment_data_dir, component_idx, 'GOCellularComponent').head(10)


# ID                                               Desc     BPval  \
5   GO:0005606                                  laminin-1 complex  0.000001   
7   GO:0043256                                    laminin complex  0.000005   
21  GO:0019031                                     viral envelope  0.000610   
48  GO:0031080                            nuclear pore outer ring  0.002727   
54  GO:0005666            DNA-directed RNA polymerase III complex  0.004041   
58  GO:0019908  nuclear cyclin-dependent protein kinase holoen...  0.004746   
72  GO:0016461                      unconventional myosin complex  0.007764   
81  GO:0005677                        chromatin silencing complex  0.010734   
85  GO:0005736              DNA-directed RNA polymerase I complex  0.011807   
93  GO:0005851  eukaryotic translation initiation factor 2B co...  0.014343   

       BFold  
5   2.325389  
7   2.148635  
21  2.839336  
48  2.255517  
54  2.228519  
58  2.188418  
72  2.067299  
81  2.111160  
85  2.494752  
93  2.793472

In [52]:
read_great_res_wrapper(enrichment_data_dir, component_idx, 'GOMolecularFunction').head(10)


# ID                                               Desc  \
3   GO:0043022                                   ribosome binding   
11  GO:0031545            peptidyl-proline 4-dioxygenase activity   
14  GO:0043208                          glycosphingolipid binding   
30  GO:0001056                        RNA polymerase III activity   
34  GO:0005007  fibroblast growth factor-activated receptor ac...   
48  GO:0032407                          MutSalpha complex binding   
60  GO:0015377                 cation:chloride symporter activity   
65  GO:0047499      calcium-independent phospholipase A2 activity   
69  GO:0032404                    mismatch repair complex binding   
74  GO:0016421                           CoA carboxylase activity   

           BPval     BFold  
3   7.584495e-08  2.351977  
11  8.138740e-06  2.466700  
14  2.462545e-05  2.495340  
30  3.860479e-04  3.131341  
34  5.214530e-04  2.250279  
48  1.512241e-03  2.575431  
60  2.143232e-03  2.088132  
65  2.303207e-03  3.545742  
69  2.498749e-03  2.155177  
74  2.932712e-03  2.029746

#### PC6 (the thrird component)


In [53]:
component_idx=6

#### what assays are driving this component?

In [54]:
np.argsort(-u_dot_d_find_assay[:, component_idx])[:5]

array([ 10, 102,  43,  35,  55])

In [55]:
u_dot_d_find_assay[np.argsort(-u_dot_d_find_assay[:, component_idx])[:5], component_idx]

array([0.18663573, 0.086168  , 0.07523161, 0.05274594, 0.05010345])

In [56]:
metadata.iloc[np.argsort(-u_dot_d_find_assay[:, component_idx])[:5], :]

sample_number antibody Genome assembly Antigen class       Antigen  \
10      ERX329621      NaN            hg19  Unclassified  Unclassified   
102     ERX329713      NaN            hg19  Unclassified  Unclassified   
43      ERX329654      NaN            hg19  Unclassified  Unclassified   
35      ERX329646      NaN            hg19  Unclassified  Unclassified   
55      ERX329666      NaN            hg19  Unclassified  Unclassified   

    Cell type class                 Cell type  \
10            Blood  Lymphoblastoid cell line   
102           Blood  Lymphoblastoid cell line   
43            Blood  Lymphoblastoid cell line   
35            Blood  Lymphoblastoid cell line   
55            Blood  Lymphoblastoid cell line   

                                 Cell type description  \
10   Tissue=blood|Lineage=mesoderm|Description=pare...   
102  Tissue=blood|Lineage=mesoderm|Description=pare...   
43   Tissue=blood|Lineage=mesoderm|Description=pare...   
35   Tissue=blood|Lineage=mesoderm|Description=pare...   
55   Tissue=blood|Lineage=mesoderm|Description=pare...   

               Processing logs  \
10   219047965,92.5,75.8,70997   
102  235003947,93.0,82.0,55388   
43    74449496,92.0,31.2,25320   
35    62384167,93.2,16.4,27131   
55    76235942,92.7,15.1,28768   

                                                 Title   ...   age  treatment  \
10   Illumina HiSeq 2000 sequencing; Coordinated ef...   ...   NaN        NaN   
102  Illumina HiSeq 2000 sequencing; Coordinated ef...   ...   NaN        NaN   
43   Illumina HiSeq 2000 sequencing; Coordinated ef...   ...   NaN        NaN   
35   Illumina HiSeq 2000 sequencing; Coordinated ef...   ...   NaN        NaN   
55   Illumina HiSeq 2000 sequencing; Coordinated ef...   ...   NaN        NaN   

    genotype lab  age.1 health state  cell_type  tissue_type  provider     sex  
10       NaN NaN    NaN          NaN    NA12891          NaN       NaN    male  
102      NaN NaN    NaN          NaN    NA19239          NaN       NaN    male  
43       NaN NaN    NaN          NaN    NA11831          NaN       NaN    male  
35       NaN NaN    NaN          NaN    NA11881          NaN       NaN    male  
55       NaN NaN    NaN          NaN    NA11840          NaN       NaN  female  

[5 rows x 24 columns]

#### what genomic bins are driving this component?

In [57]:
np.argsort(-v_dot_d_find_loci[:, component_idx])[:5]

array([211314, 347349, 167403,  12806,  12805])

In [58]:
v_dot_d_find_loci[np.argsort(-v_dot_d_find_loci[:, component_idx])[:5], component_idx]

array([9.74653295e-05, 9.42709371e-05, 9.39630245e-05, 9.37192353e-05,
       9.21795390e-05])

#### where is our loci of interest in this ranking?

In [59]:
np.sum(v_dot_d_find_loci[:, component_idx] >= v_dot_d_find_loci[genomic_bin_idx, component_idx])

8343

In [60]:
np.sum(v_dot_d_find_loci[:, component_idx] >= v_dot_d_find_loci[genomic_bin_idx, component_idx]) / v_dot_d_find_loci.shape[0]

0.021981814876390168

It's roughly on the top 2 percentile.


#### results of the enrichment analysis

In [61]:
read_great_res_wrapper(enrichment_data_dir, component_idx, 'HumanPhenotypeOntology').head(10)


# ID                             Desc         BPval     BFold
5   HP:0002633                       Vasculitis  2.124781e-09  3.254925
7   HP:0010622  Neoplasm of the skeletal system  3.157741e-09  2.848685
8   HP:0003180             Flat acetabular roof  4.132922e-09  5.580264
12  HP:0002960                     Autoimmunity  1.393192e-08  2.433056
15  HP:0002665                         Lymphoma  2.574896e-08  2.209538
16  HP:0003170    Abnormality of the acetabulum  2.993326e-08  2.925990
19  HP:0001658            Myocardial infarction  5.354662e-08  4.502097
20  HP:0000132                      Menorrhagia  5.880600e-08  4.965326
21  HP:0002135      Basal ganglia calcification  8.458347e-08  3.611860
22  HP:0002843           Abnormality of T cells  1.587267e-07  2.604574

In [62]:
read_great_res_wrapper(enrichment_data_dir, component_idx, 'GOBiologicalProcess').head(10)


# ID                                               Desc  \
1   GO:0002757     immune response-activating signal transduction   
2   GO:0050778             positive regulation of immune response   
3   GO:0002253                      activation of immune response   
9   GO:0002429  immune response-activating cell surface recept...   
15  GO:0045088               regulation of innate immune response   
16  GO:0019221                cytokine-mediated signaling pathway   
19  GO:0097190                        apoptotic signaling pathway   
23  GO:0050851        antigen receptor-mediated signaling pathway   
26  GO:0002221     pattern recognition receptor signaling pathway   
27  GO:0043122    regulation of I-kappaB kinase/NF-kappaB cascade   

           BPval     BFold  
1   8.572361e-41  2.295897  
2   4.826825e-38  2.046540  
3   2.004579e-37  2.166998  
9   5.641257e-28  2.373490  
15  3.365965e-24  2.109560  
16  5.156920e-24  2.048872  
19  1.714413e-22  2.009059  
23  7.368085e-20  2.371907  
26  3.293672e-19  2.240385  
27  4.854494e-19  2.085819

In [63]:
read_great_res_wrapper(enrichment_data_dir, component_idx, 'MGIPhenotype').head(10)


# ID                                           Desc         BPval  \
18  MP:0001828                     abnormal T cell activation  1.629343e-36   
22  MP:0001844                            autoimmune response  4.427369e-35   
23  MP:0005000                      abnormal immune tolerance  5.041626e-35   
24  MP:0005005                        abnormal self tolerance  2.471655e-34   
27  MP:0002425  altered susceptibility to autoimmune disorder  1.888043e-33   
28  MP:0005094                  abnormal T cell proliferation  4.641244e-33   
31  MP:0008217                     abnormal B cell activation  1.516460e-31   
33  MP:0002357          abnormal spleen white pulp morphology  6.913712e-31   
45  MP:0000702                           enlarged lymph nodes  2.242230e-28   
46  MP:0005153                  abnormal B cell proliferation  2.963331e-28   

       BFold  
18  2.054944  
22  2.066490  
23  2.052396  
24  2.044852  
27  2.213413  
28  2.065962  
31  2.441083  
33  2.084210  
45  2.345243  
46  2.398298

In [64]:
read_great_res_wrapper(enrichment_data_dir, component_idx, 'MGIPhenoSingleKO').head(10)


# ID                                           Desc         BPval  \
6   MP:0002459                     abnormal B cell physiology  3.216056e-49   
22  MP:0001844                            autoimmune response  3.257563e-33   
24  MP:0005000                      abnormal immune tolerance  1.527182e-32   
26  MP:0005005                        abnormal self tolerance  2.477071e-32   
29  MP:0008217                     abnormal B cell activation  7.836361e-31   
30  MP:0002425  altered susceptibility to autoimmune disorder  5.416048e-30   
34  MP:0008171              abnormal mature B cell morphology  6.807466e-28   
35  MP:0000691                                enlarged spleen  7.524606e-28   
37  MP:0001828                     abnormal T cell activation  2.567321e-27   
41  MP:0008713                        abnormal cytokine level  3.550811e-26   

       BFold  
6   2.078873  
22  2.247717  
24  2.213924  
26  2.214608  
29  2.706302  
30  2.337811  
34  2.182010  
35  2.054668  
37  2.086158  
41  2.088451

In [65]:
read_great_res_wrapper(enrichment_data_dir, component_idx, 'GOCellularComponent').head(10)


# ID                                     Desc         BPval  \
9   GO:0031094           platelet dense tubular network  2.394674e-10   
15  GO:0071682                  endocytic vesicle lumen  2.593774e-09   
20  GO:0031228              intrinsic to Golgi membrane  8.814481e-09   
22  GO:0042611                      MHC protein complex  1.732951e-08   
23  GO:0097208                   alveolar lamellar body  1.778326e-08   
26  GO:0030173               integral to Golgi membrane  6.166199e-08   
31  GO:0031095  platelet dense tubular network membrane  1.737179e-07   
33  GO:0042599                            lamellar body  2.045241e-07   
34  GO:0097342                              ripoptosome  3.640009e-07   
46  GO:0001520                        outer dense fiber  3.061652e-06   

       BFold  
9   5.336451  
15  5.427595  
20  2.196942  
22  4.831563  
23  8.062619  
26  2.153514  
31  4.607586  
33  5.946830  
34  8.765341  
46  6.119236

In [66]:
read_great_res_wrapper(enrichment_data_dir, component_idx, 'GOMolecularFunction').head(10)


# ID                                            Desc         BPval  \
1   GO:0005126                       cytokine receptor binding  2.201581e-17   
6   GO:0002020                                protease binding  1.271400e-10   
10  GO:0016835                    carbon-oxygen lyase activity  2.251880e-09   
11  GO:0051861                              glycolipid binding  4.785386e-09   
16  GO:0032393                   MHC class I receptor activity  6.452181e-08   
17  GO:0005086  ARF guanyl-nucleotide exchange factor activity  6.820259e-08   
19  GO:0016836                            hydro-lyase activity  1.034720e-07   
20  GO:0045236                 CXCR chemokine receptor binding  1.167359e-07   
25  GO:0003755    peptidyl-prolyl cis-trans isomerase activity  2.664467e-07   
27  GO:0046625                            sphingolipid binding  3.306067e-07   

       BFold  
1   2.035422  
6   2.558850  
10  2.474315  
11  4.060440  
16  7.203130  
17  3.116550  
19  2.542034  
20  4.494398  
25  2.421004  
27  3.698747